coding_standards.md

```diff
@@ -0,0 +1,95 @@
+# Coding standards
+
+These are common standards that apply to coding.
+
+Mostly these follow common [programming principles](https://en.wikipedia.org/wiki/Category:Programming_principles)
+
+## Rationale
+
+- Ensure consistency across coding standards for different languages
+
+## Standards
+
+### All code is open
+
+Whilst we acknowledge that there are cases where it may be necessary for code to not be open, this standard ensure that those cases are amanged as exceptions.
+
+### All coding is done on a branch
+
+No code changes are made directly to the master branch.
+This reflects the 'golden' state of the branch and mirrors the general standard that changes are never made directly to production services.
+
+### All code is checked for correct coding style
+
+Language-dependent
+
+### All branches have CI
+
+Covering code style, security checking, dependency checking, build and unit test
+
+### Unit test coverage is at least 90%
+
+All changes must have sufficient coverage and the overall total coverage must not decrease.
```

Comment:
This is a bit difficult to apply across all developments.  For example a .Net MVC project on creation will have a lot of microsoft boiler plate not covered by unit tests.  For smaller projects hitting 90% is not feasible.  Perhaps rewording to state x% of written code (ie code we have written in Defra as opposed by someone else).

I think given the nature of some developments 90% may still be a bit high.  What evidence have we used to decide this figure?

Reply:
The rationale for having 90% is a recognition that 100% is generally not practical for most code and we have managed to sustain 90% or above on all our newer code.

As far as boilerplate is concerned, I would expect that either it would get exercised by other unit tests anyway or it is performing framework-y functions that you wouldn't expect to be invoking as part of unit testing, so there may be some sense in define them as specific exclusions in whatever testing framework you're using.

We should probably discuss that further and look at specific examples - it might be that we can articulate something in terms of specific standards for particular languages or frameworks.

But I think a common, general "90% coverage" is the right standard to adopt overall.

Reply:
For .Net most of the boilerplate code I refer to is standard code Microsoft apply to project templates.  For example setting up default routing of MVC projects, server start up and shut down code etc.

For the majority of projects you wouldn't be changing this pre-written code but it does count towards your code coverage as there will be no unit tests covering it.  I wouldn't expect developers to spend time writing unit tests to test code written by Microsoft, just like I wouldn't expect them to write tests for third party libraries they've added.

For smaller projects it would be very difficult to obtain 90%.

For open source projects where pretty much all code is written from scratch 90% is more achievable.

I agree some examples would probably help demonstrate.

Reply:
Perhaps we make an exception for C#, in which case we just need to agree what would be appropriate. And we could just go with that for now.

I do think this is something that is worth looking into a bit deeper in future, as I'm sure someone has tackled this problem.

Reply:
I don't think C# is the only exception here. I agree that this is a good target but perhaps we should consider what is appropriate to the language / task / framework / risk of the code failing.